File: mediator/src/main/java/com/iluwatar/mediator/PartyMemberBase.java
package com.iluwatar.mediator;

/**
 * 
 * Abstract base class for party members.
 *
 */
public abstract class PartyMemberBase implements PartyMember {

	protected Party party;

	@Override
	public void joinedParty(Party party) {
		System.out.println(this + " joins the party");
		this.party = party;
	}

	@Override
	public void partyAction(Action action) {
		System.out.println(this + " " + action.getDescription());
	}

	@Override
	public void act(Action action) {
		if (party != null) {
			System.out.println(this + " " + action.toString());
			party.act(this, action);
		}
	}

	@Override
	public abstract String toString();

}

File: mediator/src/main/kotlin/com/iluwatar/mediator/PartyMemberBase.kt
/*
 * This project is licensed under the MIT license. Module model-view-viewmodel is using ZK framework licensed under LGPL (see lgpl-3.0.txt).
 *
 * The MIT License
 * Copyright © 2014-2022 Ilkka Seppälä
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.iluwatar.mediator

// ABOUTME: Abstract base class for party members providing shared behavior.
// ABOUTME: Handles joining a party, reacting to party actions, and performing actions via the mediator.

import io.github.oshai.kotlinlogging.KotlinLogging

private val logger = KotlinLogging.logger {}

/** Abstract base class for party members. */
abstract class PartyMemberBase : PartyMember {

    protected var party: Party? = null

    override fun joinedParty(party: Party) {
        logger.info { "$this joins the party" }
        this.party = party
    }

    override fun partyAction(action: Action) {
        logger.info { "$this ${action.description}" }
    }

    override fun act(action: Action) {
        party?.let {
            logger.info { "$this $action" }
            it.act(this, action)
        }
    }

    abstract override fun toString(): String
}
